File: Week 5/1. Service/ros2_ws/src/pkg_cpp_py/scripts/add_two_ints_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
import sys

class Client(Node):
    def __init__(self):
        super().__init__("add_two_ints")
        self.client = self.create_client(AddTwoInts, "add_two_ints")

        while not self.client.wait_for_service(1.0):
            self.get_logger().warn("Server not available, waiting...")

        self.request = AddTwoInts.Request()

    def send_request(self, a, b):
        self.request.a = a
        self.request.b = b
        self.future = self.client.call_async(self.request)
        rclpy.spin_until_future_complete(self, self.future)
        return self.future.result()

def main(args=None):
    rclpy.init(args=args)
    node = Client()
    response = node.send_request(int(sys.argv[1]), int(sys.argv[2]))
    node.get_logger().info(f'{int(sys.argv[1])} + { int(sys.argv[2])} = {response.sum}')
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Week 5/1. Service/ros2_ws/src/pkg_cpp_py/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using namespace std::chrono_literals;

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);

    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client");
    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    // check for server
    while(!client->wait_for_service(1s)){
        RCLCPP_WARN(node->get_logger(), "Server not available, waiting...");
    }

    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
    request->a = atoll(argv[1]);
    request->b = atoll(argv[2]);

    auto future = client->async_send_request(request);

    // wait for result
    if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS){
        RCLCPP_INFO(node->get_logger(), "The sum of %d and %d is %d", request->a, request->b, future.get()->sum);
    } else {
        RCLCPP_ERROR(node->get_logger(), "Ooops! failed to call service.");
    }

    return 0;
}